osu.Game/Rulesets/Scoring/ScoreProcessor.cs

```diff
@@ -445,6 +453,32 @@ public virtual void PopulateScore(ScoreInfo score)
             score.TotalScore = (long)Math.Round(ComputeFinalScore(ScoringMode.Standardised, score));
         }
 
+        /// <summary>
+        /// Populates the given score with remaining statistics as "missed" and marks it with <see cref="ScoreRank.F"/> rank.
+        /// </summary>
+        public void FailScore(ScoreInfo score)
+        {
+            if (Rank.Value == ScoreRank.F)
+                return;
+
+            Rank.Value = ScoreRank.F;
+            Rank.Disabled = true;
+
+            Debug.Assert(maximumResultCounts != null);
+
+            if (maximumResultCounts.TryGetValue(HitResult.LargeTickHit, out int maximumLargeTick))
+                scoreResultCounts[HitResult.LargeTickMiss] = maximumLargeTick - scoreResultCounts.GetValueOrDefault(HitResult.LargeTickHit);
+
+            if (maximumResultCounts.TryGetValue(HitResult.SmallTickHit, out int maximumSmallTick))
+                scoreResultCounts[HitResult.SmallTickMiss] = maximumSmallTick - scoreResultCounts.GetValueOrDefault(HitResult.SmallTickHit);
+
+            int maximumBasic = maximumResultCounts.SingleOrDefault(kvp => kvp.Key.IsBasic()).Value;
+            int currentBasic = scoreResultCounts.Where(kvp => kvp.Key.IsBasic() && kvp.Key != HitResult.Miss).Sum(kvp => kvp.Value);
+            scoreResultCounts[HitResult.Miss] = maximumBasic - currentBasic;
+
```

Comment:
Count of `IgnoreHit` and `IgnoreMiss` should also be written out to ensure `JudgedHits` can be calculated. See https://github.com/ppy/osu/pull/16722 for more info.